### PR TITLE
ADD: tests to google oauth 

### PR DESCRIPTION
**Pull request by MennaTullahTaha**
The original code submitted in the branch feature/oauth was not testable. 

I refactored the resolver into a helper file in /packages/src/backend/plugins/plugins_helper/googleAuthResolver.ts and exported the resolver to make it testable. 

The tests are written in index.test.ts.

### CODE REVIEW

File: packages/backend/src/index.test.ts [hunk 1]
```diff
@@ -1,8 +1,165 @@
...
+      },
+      signInWithCatalogUser: function (
+        query: AuthResolverCatalogUserQuery,
+      ): Promise<BackstageSignInResult> {
+        throw new Error('Function not implemented.');
+      },
+    };
+
+    await expect(resolverResult(mockSignInInfo, mockContext)).rejects.toThrow(
+      `Login failed, '${mockProfile.email}' does not belong to the expected domain`,
```
[the remark this answers is not in the document]
Comment (AGiljanovic) [reply packages/backend/src/index.test.ts:130]: I'd rather do something like:
throw new Error("Login failed due to an incorrect email domain.");

File: packages/backend/src/plugins/plugins_helper/googleAuthResolver.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+  if (!profile.email || !regexp.test(profile.email)) {
+    throw new Error(
+      'Login failed, user profile does not contain a valid email',
+    );
+  }
+
+  const [localPart, domain] = profile.email.split('@');
+
+  if (domain !== 'code.berlin') {
```
Comment (AGiljanovic) [inline packages/backend/src/plugins/plugins_helper/googleAuthResolver.ts:25]: I'd consider switching this into an .env var, just for the sake of minimising the chances of exposing any data

File: packages/backend/src/index.test.ts [hunk 3]
```diff
@@ -1,8 +1,165 @@
...
+      },
+    };
+
+    await expect(resolverResult(mockSignInInfo, mockContext)).rejects.toThrow(
+      'Login failed, user profile does not contain a valid email',
+    );
+  });
+
+  it('should throw an exception for valid email address with incorrect domain', async () => {
+    const mockProfile: ProfileInfo = {
```
Comment (AGiljanovic) [inline packages/backend/src/index.test.ts:100]: Repetitive code, use `beforeEach` rather than copy pasting.

File: packages/backend/src/index.test.ts [hunk 4]
```diff
@@ -1,8 +1,165 @@
...
+      displayName: 'John Doe',
+      picture: 'https://example.com/avatar.jpg',
+    };
+
+    const mockSignInInfo: SignInInfo<any> = {
+      profile: mockProfile,
+      result: {}, // Placeholder for the authentication result
+    };
+
+    const mockContext: AuthResolverContext = {
```
Comment (AGiljanovic) [inline packages/backend/src/index.test.ts:111]: Repetitive code, use `beforeEach` rather than copy pasting.

File: packages/backend/src/index.test.ts [hunk 5]
```diff
@@ -1,8 +1,165 @@
...
+  });
+
+  it('should throw an exception for valid email address with incorrect domain', async () => {
+    const mockProfile: ProfileInfo = {
+      email: 'user@example.com',
+      displayName: 'John Doe',
+      picture: 'https://example.com/avatar.jpg',
+    };
+
+    const mockSignInInfo: SignInInfo<any> = {
```
Comment (AGiljanovic) [inline packages/backend/src/index.test.ts:106]: Repetitive code, use beforeEach rather than copy pasting.

File: packages/backend/src/index.test.ts [hunk 6]
```diff
@@ -1,8 +1,165 @@
...
+import { Entity } from '@backstage/catalog-model';
 
 describe('test', () => {
   it('unbreaks the test runner', () => {
     const unbreaker = {} as PluginEnvironment;
     expect(unbreaker).toBeTruthy();
   });
 });
+
+describe('providers.google.create.signIn.resolver logic', () => {
```
Comment (AGiljanovic) [inline packages/backend/src/index.test.ts:20]: to fix the repetitive logic; use a `beforeEach` stead
```
 let mockProfile, mockSignInInfo, mockContext;

 beforeEach(() => {
    mockProfile = { displayName: 'John Doe', picture: 'https://example.com/avatar.jpg' };
    mockSignInInfo = { profile: mockProfile, result: {} };
    mockContext = {
      issueToken: jest.fn().mockResolvedValue({ token: 'fake token' }),
      findCatalogUser: jest.fn(),
      signInWithCatalogUser: jest.fn(),
    };
  });
```

Comment (AGiljanovic) [reply packages/backend/src/index.test.ts:20]: This way the blocks should run before each one of the tests, with the variables set up fresh for each test case, so you've also less redundancy in the code and a more centralized setup that's easier to manage later down the line.

**Review by AGiljanovic — CHANGES_REQUESTED** [review]
I would also just add in general validation testing against the different email formats, including edge cases and potential security attacks :)

**Review by AGiljanovic — APPROVED** [review]
For future reference, stick to the naming convention, otherwise:

LGTM
